### PR TITLE
[Feature] API exception handle project delete with components

### LINKED ISSUE #3969: [bug] DELETE project with components through API throws 500 error
<h1> Steps </h1>

1. On staging2 create a node, and then create a component for the node.
2.  Via the API, attempt to delete the first node.
screen shot 2015-08-06 at 3 17 36 pm

<h1> Expected </h1>

Maybe a 405 Error Method Not Allowed with error detail explaining "Any child components must be deleted prior to deleting this project."

<h1> Actual </h1>

A 500 error is returned.  This exception needs to be handled properly.
screen shot 2015-08-06 at 3 17 51 pm

### PR DESCRIPTION
**Pull request by billyhunt**
Fixes #3969

Purpose
-------

Trying to delete a node that has children is not allowed.  Currently, the API returns a 500 error.  After the change, it will return a 403 response with the message "Any child components must be deleted prior to deleting this project." 

Changes
------------

Within the NodeDetail class, the perform_destroy method now has logic to test if the the node has children.  If so, an exception is raised.

### Before

screen shot 2015-08-12 at 4 33 15 pm

### After

screen shot 2015-08-12 at 4 33 30 pm

### CODE REVIEW

File: api/nodes/views.py [hunk 1]
```diff
@@ -109,8 +109,13 @@ def perform_destroy(self, instance):
         user = self.request.user
         auth = Auth(user)
         node = self.get_object()
-        node.remove_node(auth=auth)
-        node.save()
```
Comment (brianjgeiger) [inline api/nodes/views.py]: Does the node model not raise an exception here? I presume that's what's causing the 500. Ideally, we should just catch the exception and return the 403 if it happens.

**Comment by brianjgeiger** [comment]
Perfect. Ready for next pass.

File: api/nodes/views.py [hunk 2]
```diff
@@ -109,7 +110,11 @@ def perform_destroy(self, instance):
         user = self.request.user
         auth = Auth(user)
         node = self.get_object()
-        node.remove_node(auth=auth)
+        # If there are no children, then allow the node to be removed, otherwise give a 403 error.
+        try:
+            node.remove_node(auth=auth)
+        except NodeStateError:
+            raise PermissionDenied()
```
Comment (sloria) [inline api/nodes/views.py]: I'm not sure this is the correct response to send if a `NodeStateError` is raised. Take a look at the `Node#remove_node` method and see what conditions raise `NodeStateError`. It's also possible for a `PermissionsError` to be raised, which should also be handled.

**Comment by sloria** [comment]
Pass finished. :vibration_mode:

**Comment by sloria** [comment]
The fix for this is included in https://github.com/CenterForOpenScience/osf.io/pull/4254 . Thanks for your work on this!

**Pull request closed without merging**